### PR TITLE
[System.ServiceModel] Fix timeout defaulting to 0 seconds

### PR DESCRIPTION
**Pull request by alexanderkyte**

### CODE REVIEW

File: mcs/class/System.ServiceModel/System.ServiceModel.Description/MetadataExchangeClient.cs [hunk 1]
```diff
@@ -149,7 +160,7 @@ internal MetadataSet GetMetadataInternal (EndpointAddress address, MetadataExcha
 			var asClientChannel = proxy as IClientChannel;
 			if (asClientChannel == null)
 				throw new InvalidOperationException ("The channel factory must return an IClientChannel implementation");
-			asClientChannel.OperationTimeout = OperationTimeout;
+			asClientChannel.OperationTimeout = this.OperationTimeout;
```
Comment (alexrp) [inline mcs/class/System.ServiceModel/System.ServiceModel.Description/MetadataExchangeClient.cs]: why this change?

File: mcs/class/System.ServiceModel/System.ServiceModel.Description/MetadataExchangeClient.cs [hunk 2]
```diff
@@ -88,7 +89,17 @@ public MetadataExchangeClient (Uri address, MetadataExchangeClientMode mode)
 		[MonoTODO]
 		public int MaximumResolvedReferences { get; set; }
 
-		public TimeSpan OperationTimeout { get; set; }
+		public TimeSpan OperationTimeout {
+			get {
+				if (!this.operation_timeout.HasValue) {
+					this.operation_timeout = TimeSpan.FromMinutes (1);
```
Comment (alexrp) [inline mcs/class/System.ServiceModel/System.ServiceModel.Description/MetadataExchangeClient.cs]: any particular reason for a 1 minute default? would be good to put in a comment

File: mcs/class/System.ServiceModel/System.ServiceModel/ClientRuntimeChannel.cs [hunk 3]
```diff
@@ -322,8 +321,19 @@ public void Dispose ()
...
 		}
 
-		[MonoTODO]
-		public TimeSpan OperationTimeout { get; set; }
+		TimeSpan? operation_timeout = null;
+
+		public TimeSpan OperationTimeout {
+			get {
+				if (!this.operation_timeout.HasValue) {
+					this.operation_timeout = TimeSpan.FromMinutes (1);
```
Comment (alexrp) [inline mcs/class/System.ServiceModel/System.ServiceModel/ClientRuntimeChannel.cs]: ditto

File: mcs/class/System.ServiceModel/System.ServiceModel.Description/MetadataExchangeClient.cs (hunk 2 above)
Comment (alexanderkyte) [inline mcs/class/System.ServiceModel/System.ServiceModel.Description/MetadataExchangeClient.cs]: No idea why it's the default, but it was set elsewhere. I should move it out into a #define.

Comment (akoeplinger) [inline mcs/class/System.ServiceModel/System.ServiceModel.Description/MetadataExchangeClient.cs]: [MSDN](https://msdn.microsoft.com/en-us/library/system.servicemodel.description.metadataexchangeclient.operationtimeout(v=vs.100).aspx) lists the default as 1 minute on .NET

**Comment by alexanderkyte** [comment]
Rebuild last

**Comment by alexanderkyte** [comment]
build

**Comment by migueldeicaza** [comment]
build

**Comment by alexrp** [comment]
Well, I'm no expert on WCF, but this looks sensible to me.

@migueldeicaza what do you think?

**Comment by alexrp** [comment]
build

**Comment by alexanderkyte** [comment]
I'm seeing clean on x86/amd64 and on arm I see

```
* Assertion at mini-arm.c:3781, condition `pdata.found == 1' not met
```

which is an intermittent failure that I've been seeing on most PRs on arm.

File: mcs/class/System.ServiceModel/System.ServiceModel/ClientRuntimeChannel.cs [hunk 4]
```diff
@@ -322,8 +321,19 @@ public void Dispose ()
 			}
 		}
 
-		[MonoTODO]
-		public TimeSpan OperationTimeout { get; set; }
+		TimeSpan? operation_timeout = null;
```
Comment (kumpera) [inline mcs/class/System.ServiceModel/System.ServiceModel/ClientRuntimeChannel.cs]: This should go to the top with all the other fields.

**Comment by kumpera** [comment]
LGTM

**Comment by alexanderkyte** [comment]
Moved it up.
